### PR TITLE
DRIVERS-1652: Clarify events.json and result.json produced by workload executor

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/DRIVERS-1652

This should also address [DRIVERS-1604](https://jira.mongodb.org/browse/DRIVERS-1604).

### CODE REVIEW

File: astrolabe/utils.py
Comment (jmikola) [inline astrolabe/utils.py:267]: I found this was helpful for debugging, since this file is never reported elsewhere and also does not get saved as a build artifact like `events.json`. That said, saving `events.json` isn't very helpful for the validation tests since only the final copy of it will be persisted.

File: astrolabe/utils.py
Comment (jmikola) [inline astrolabe/utils.py:271]: The workload executor is always expected to produce a (presumably valid) `results.json` file, so I removed `_PLACEHOLDER_EXECUTION_STATISTICS` in favor of an explicit error. This also guarantees to the caller that any returned stats will have originated from a real file produced by the workload executor.

File: astrolabe/validator.py
Comment (jmikola) [inline astrolabe/validator.py:75]: `DriverWorkloadSubprocessRunner.stop()` actually never raised a `TimeoutExpired` error. It was caught internally and re-thrown as a `WorkloadExecutorError`.

File: astrolabe/validator.py
Comment (jmikola) [inline astrolabe/validator.py:104]: This is catching a potential exception from `subprocess.stop()` (re-thrown TimeoutExpired) and `subprocess.read_stats()`. In either case, we don't have stats to report so we're failing the test explicitly instead of using placeholder stats.

File: astrolabe/validator.py
Comment (jmikola) [inline astrolabe/validator.py:180]: I was originally asserting that this was a float but later realized that was unreliable. For example, Java produces `1.618408012349E9` as a numeric literal in its `events.json`, which `yaml.safe_load` parses as a string.

File: astrolabe/validator.py
Comment (jmikola) [inline astrolabe/validator.py:167]: I intend to come back to this comment, as @p-mongo convinced me to leave in the code below that asserts `events` has at least one object of the types: Connection, Pool, and Command.

While I could trivially check an environment variable from within `validator.py` and disable those checks for libmongoc and PHP (neither of which have CMAP), I feel like there's a more idiomatic approach to handle this through a command line option. Perhaps @prashantmital can chime in with some advice here.

I'm currently working around this in PHPLIB (#115) by injecting fake CMAP events into the array, so it's not urgent. We can easily punt this to a separate ticket to avoid holding up this PR.

File: astrolabe/validator.py
Comment (jmikola) [inline astrolabe/validator.py:229]: Similar to `observedAt` for events, I chose not to assert the type of `time` here.

File: docs/source/spec-test-format.rst
Comment (jmikola) [inline docs/source/spec-test-format.rst:146]: IIRC, when @p-mongo was implementing the unified test runner changes we concluded that the actual scenario tests should only use `storeErrorsAsEntity` so that Astrolabe could reliably consult `errors` and `numErrors` without having to worry about how drivers might distinguish errors and failures (assuming they do at all).

Since we still test `storeFailuresAsEntity` in the validation tests, I thought that deserved a mention here.

File: docs/source/spec-workload-executor.rst
Comment (jmikola) [inline docs/source/spec-workload-executor.rst:75]: Initializing all of the output fields for the JSON files up front avoids the ambiguity about what happens when the test runner fails. Now, we always have _some_ value to use for all of these fields, and the error/failure counts are always calculated from the corresponding arrays.

File: docs/source/spec-workload-executor.rst
Comment (jmikola) [inline docs/source/spec-workload-executor.rst:89]: I chose to use "append" whenever we talk about modifying the arrays to avoid any race condition in overwriting the `errors` array. That could happen if the workload executor caught a propagated error from the test runner and also fetched errors from the entity map. For instance, PHPLIB's test runner always invokes the entity map callback (used by our workload executor to fetch values) after a test, even if it ultimately throws an exception. We invoke the callback from a `finally` block.

While it's unlikely that a single test would _both_ report errors via an entity _and_ propagate an error, it's theoretically possible (e.g. `loop` that catches errors and then some error in a separate operation after the `loop`).

File: astrolabe/validator.py
Comment (p-mongo) [inline astrolabe/validator.py:78]: Could you clarify why this was removed or if this logic was moved to another place?

Comment (jmikola) [reply astrolabe/validator.py:78]: See: https://github.com/mongodb-labs/drivers-atlas-testing/pull/120/files#r613277944

**Comment by jmikola** [comment]
Patch build https://spruce.mongodb.com/version/6076f61b9ccd4e55a2324d16/tasks

### Java failures:

```
[2021/04/14 14:10:19.232] ======================================================================
 [2021/04/14 14:10:19.232] FAIL: test_num_errors (astrolabe.validator.ValidateWorkloadExecutor)
 [2021/04/14 14:10:19.232] ----------------------------------------------------------------------
 [2021/04/14 14:10:19.232] Traceback (most recent call last):
 [2021/04/14 14:10:19.232]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 327, in test_num_errors
 [2021/04/14 14:10:19.232]     self.assert_events(hasEvents=False, hasErrors=True, hasFailures=False)
 [2021/04/14 14:10:19.232]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 226, in assert_events
 [2021/04/14 14:10:19.232]     self.fail("The workload executor recorded no errors but some were expected")
 [2021/04/14 14:10:19.232] AssertionError: The workload executor recorded no errors but some were expected
 [2021/04/14 14:10:19.232] ======================================================================
 [2021/04/14 14:10:19.232] FAIL: test_num_errors_not_captured (astrolabe.validator.ValidateWorkloadExecutor)
 [2021/04/14 14:10:19.232] ----------------------------------------------------------------------
 [2021/04/14 14:10:19.232] Traceback (most recent call last):
 [2021/04/14 14:10:19.232]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/utils.py", line 238, in stop
 [2021/04/14 14:10:19.232]     os.killpg(self.workload_subprocess.pid, signal.SIGINT)
 [2021/04/14 14:10:19.232] ProcessLookupError: [Errno 3] No such process
 [2021/04/14 14:10:19.232] During handling of the above exception, another exception occurred:
 [2021/04/14 14:10:19.232] Traceback (most recent call last):
 [2021/04/14 14:10:19.232]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 100, in run_test_expecting_error
 [2021/04/14 14:10:19.232]     stats = subprocess.stop()
 [2021/04/14 14:10:19.232]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/utils.py", line 242, in stop
 [2021/04/14 14:10:19.232]     raise PrematureExitError("Could not request termination of workload executor, possibly because the workload executor exited prematurely: %s" % exc)
 [2021/04/14 14:10:19.232] astrolabe.exceptions.PrematureExitError: Could not request termination of workload executor, possibly because the workload executor exited prematurely: [Errno 3] No such process
 [2021/04/14 14:10:19.232] During handling of the above exception, another exception occurred:
 [2021/04/14 14:10:19.232] Traceback (most recent call last):
 [2021/04/14 14:10:19.232]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/utils.py", line 265, in read_stats
 [2021/04/14 14:10:19.232]     with open(self.sentinel, 'r') as fp:
 [2021/04/14 14:10:19.232] FileNotFoundError: [Errno 2] No such file or directory: '/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/results.json'
 [2021/04/14 14:10:19.233] During handling of the above exception, another exception occurred:
 [2021/04/14 14:10:19.233] Traceback (most recent call last):
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 102, in run_test_expecting_error
 [2021/04/14 14:10:19.233]     stats = subprocess.read_stats()
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/utils.py", line 271, in read_stats
 [2021/04/14 14:10:19.233]     raise WorkloadExecutorError("The workload executor did not write a results.json file in the expected location")
 [2021/04/14 14:10:19.233] astrolabe.exceptions.WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:10:19.233] During handling of the above exception, another exception occurred:
 [2021/04/14 14:10:19.233] Traceback (most recent call last):
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 334, in test_num_errors_not_captured
 [2021/04/14 14:10:19.233]     stats = self.run_test_expecting_error(driver_workload)
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 104, in run_test_expecting_error
 [2021/04/14 14:10:19.233]     self.fail("WorkloadExecutorError: %s" % exc)
 [2021/04/14 14:10:19.233] AssertionError: WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:10:19.233] ======================================================================
 [2021/04/14 14:10:19.233] FAIL: test_num_failures_not_captured (astrolabe.validator.ValidateWorkloadExecutor)
 [2021/04/14 14:10:19.233] ----------------------------------------------------------------------
 [2021/04/14 14:10:19.233] Traceback (most recent call last):
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/utils.py", line 238, in stop
 [2021/04/14 14:10:19.233]     os.killpg(self.workload_subprocess.pid, signal.SIGINT)
 [2021/04/14 14:10:19.233] ProcessLookupError: [Errno 3] No such process
 [2021/04/14 14:10:19.233] During handling of the above exception, another exception occurred:
 [2021/04/14 14:10:19.233] Traceback (most recent call last):
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 100, in run_test_expecting_error
 [2021/04/14 14:10:19.233]     stats = subprocess.stop()
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/utils.py", line 242, in stop
 [2021/04/14 14:10:19.233]     raise PrematureExitError("Could not request termination of workload executor, possibly because the workload executor exited prematurely: %s" % exc)
 [2021/04/14 14:10:19.233] astrolabe.exceptions.PrematureExitError: Could not request termination of workload executor, possibly because the workload executor exited prematurely: [Errno 3] No such process
 [2021/04/14 14:10:19.233] During handling of the above exception, another exception occurred:
 [2021/04/14 14:10:19.233] Traceback (most recent call last):
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/utils.py", line 265, in read_stats
 [2021/04/14 14:10:19.233]     with open(self.sentinel, 'r') as fp:
 [2021/04/14 14:10:19.233] FileNotFoundError: [Errno 2] No such file or directory: '/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/results.json'
 [2021/04/14 14:10:19.233] During handling of the above exception, another exception occurred:
 [2021/04/14 14:10:19.233] Traceback (most recent call last):
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 102, in run_test_expecting_error
 [2021/04/14 14:10:19.233]     stats = subprocess.read_stats()
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/utils.py", line 271, in read_stats
 [2021/04/14 14:10:19.233]     raise WorkloadExecutorError("The workload executor did not write a results.json file in the expected location")
 [2021/04/14 14:10:19.233] astrolabe.exceptions.WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:10:19.233] During handling of the above exception, another exception occurred:
 [2021/04/14 14:10:19.233] Traceback (most recent call last):
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 429, in test_num_failures_not_captured
 [2021/04/14 14:10:19.233]     stats = self.run_test_expecting_error(driver_workload)
 [2021/04/14 14:10:19.233]   File "/data/mci/0c04e3530daf482a5afc88e155703886/astrolabe-src/astrolabe/validator.py", line 104, in run_test_expecting_error
 [2021/04/14 14:10:19.233]     self.fail("WorkloadExecutorError: %s" % exc)
 [2021/04/14 14:10:19.233] AssertionError: WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:10:19.233] ----------------------------------------------------------------------
 [2021/04/14 14:10:19.233] Ran 7 tests in 45.471s
 [2021/04/14 14:10:19.233] FAILED (failures=3)
```

### Ruby failures

```
 [2021/04/14 14:11:49.747] ======================================================================
 [2021/04/14 14:11:49.747] FAIL: test_num_errors (astrolabe.validator.ValidateWorkloadExecutor)
 [2021/04/14 14:11:49.747] ----------------------------------------------------------------------
 [2021/04/14 14:11:49.747] Traceback (most recent call last):
 [2021/04/14 14:11:49.747]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 327, in test_num_errors
 [2021/04/14 14:11:49.747]     self.assert_events(hasEvents=False, hasErrors=True, hasFailures=False)
 [2021/04/14 14:11:49.747]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 230, in assert_events
 [2021/04/14 14:11:49.747]     self.fail("The workload executor didn't record error as expected: {}".format(error))
 [2021/04/14 14:11:49.747] AssertionError: The workload executor didn't record error as expected: {'error': "NoMethodError: undefined method `does_not_exist' for #<Unified::Test:0x0000559322df13e8>", 'observedAt': 1618409478.0300002}
 [2021/04/14 14:11:49.747] ======================================================================
 [2021/04/14 14:11:49.747] FAIL: test_num_errors_as_failures (astrolabe.validator.ValidateWorkloadExecutor)
 [2021/04/14 14:11:49.747] ----------------------------------------------------------------------
 [2021/04/14 14:11:49.747] Traceback (most recent call last):
 [2021/04/14 14:11:49.747]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 265, in read_stats
 [2021/04/14 14:11:49.747]     with open(self.sentinel, 'r') as fp:
 [2021/04/14 14:11:49.747] FileNotFoundError: [Errno 2] No such file or directory: '/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/results.json'
 [2021/04/14 14:11:49.747] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.747] Traceback (most recent call last):
 [2021/04/14 14:11:49.747]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 73, in run_test
 [2021/04/14 14:11:49.747]     stats = subprocess.stop()
 [2021/04/14 14:11:49.747]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 260, in stop
 [2021/04/14 14:11:49.747]     return self.read_stats()
 [2021/04/14 14:11:49.747] astrolabe.exceptions.WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:11:49.747] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.747] Traceback (most recent call last):
 [2021/04/14 14:11:49.747]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 363, in test_num_errors_as_failures
 [2021/04/14 14:11:49.747]     stats = self.run_test(driver_workload)
 [2021/04/14 14:11:49.747]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 75, in run_test
 [2021/04/14 14:11:49.747]     self.fail("WorkloadExecutorError: %s" % exc)
 [2021/04/14 14:11:49.747] AssertionError: WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:11:49.747] ======================================================================
 [2021/04/14 14:11:49.747] FAIL: test_num_errors_not_captured (astrolabe.validator.ValidateWorkloadExecutor)
 [2021/04/14 14:11:49.748] ----------------------------------------------------------------------
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 238, in stop
 [2021/04/14 14:11:49.748]     os.killpg(self.workload_subprocess.pid, signal.SIGINT)
 [2021/04/14 14:11:49.748] ProcessLookupError: [Errno 3] No such process
 [2021/04/14 14:11:49.748] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 100, in run_test_expecting_error
 [2021/04/14 14:11:49.748]     stats = subprocess.stop()
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 242, in stop
 [2021/04/14 14:11:49.748]     raise PrematureExitError("Could not request termination of workload executor, possibly because the workload executor exited prematurely: %s" % exc)
 [2021/04/14 14:11:49.748] astrolabe.exceptions.PrematureExitError: Could not request termination of workload executor, possibly because the workload executor exited prematurely: [Errno 3] No such process
 [2021/04/14 14:11:49.748] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 265, in read_stats
 [2021/04/14 14:11:49.748]     with open(self.sentinel, 'r') as fp:
 [2021/04/14 14:11:49.748] FileNotFoundError: [Errno 2] No such file or directory: '/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/results.json'
 [2021/04/14 14:11:49.748] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 102, in run_test_expecting_error
 [2021/04/14 14:11:49.748]     stats = subprocess.read_stats()
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 271, in read_stats
 [2021/04/14 14:11:49.748]     raise WorkloadExecutorError("The workload executor did not write a results.json file in the expected location")
 [2021/04/14 14:11:49.748] astrolabe.exceptions.WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:11:49.748] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 334, in test_num_errors_not_captured
 [2021/04/14 14:11:49.748]     stats = self.run_test_expecting_error(driver_workload)
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 104, in run_test_expecting_error
 [2021/04/14 14:11:49.748]     self.fail("WorkloadExecutorError: %s" % exc)
 [2021/04/14 14:11:49.748] AssertionError: WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:11:49.748] ======================================================================
 [2021/04/14 14:11:49.748] FAIL: test_num_failures_not_captured (astrolabe.validator.ValidateWorkloadExecutor)
 [2021/04/14 14:11:49.748] ----------------------------------------------------------------------
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 238, in stop
 [2021/04/14 14:11:49.748]     os.killpg(self.workload_subprocess.pid, signal.SIGINT)
 [2021/04/14 14:11:49.748] ProcessLookupError: [Errno 3] No such process
 [2021/04/14 14:11:49.748] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 100, in run_test_expecting_error
 [2021/04/14 14:11:49.748]     stats = subprocess.stop()
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 242, in stop
 [2021/04/14 14:11:49.748]     raise PrematureExitError("Could not request termination of workload executor, possibly because the workload executor exited prematurely: %s" % exc)
 [2021/04/14 14:11:49.748] astrolabe.exceptions.PrematureExitError: Could not request termination of workload executor, possibly because the workload executor exited prematurely: [Errno 3] No such process
 [2021/04/14 14:11:49.748] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 265, in read_stats
 [2021/04/14 14:11:49.748]     with open(self.sentinel, 'r') as fp:
 [2021/04/14 14:11:49.748] FileNotFoundError: [Errno 2] No such file or directory: '/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/results.json'
 [2021/04/14 14:11:49.748] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 102, in run_test_expecting_error
 [2021/04/14 14:11:49.748]     stats = subprocess.read_stats()
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/utils.py", line 271, in read_stats
 [2021/04/14 14:11:49.748]     raise WorkloadExecutorError("The workload executor did not write a results.json file in the expected location")
 [2021/04/14 14:11:49.748] astrolabe.exceptions.WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:11:49.748] During handling of the above exception, another exception occurred:
 [2021/04/14 14:11:49.748] Traceback (most recent call last):
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 429, in test_num_failures_not_captured
 [2021/04/14 14:11:49.748]     stats = self.run_test_expecting_error(driver_workload)
 [2021/04/14 14:11:49.748]   File "/data/mci/3b4344804934c30b581d8864e61a3af9/astrolabe-src/astrolabe/validator.py", line 104, in run_test_expecting_error
 [2021/04/14 14:11:49.748]     self.fail("WorkloadExecutorError: %s" % exc)
 [2021/04/14 14:11:49.748] AssertionError: WorkloadExecutorError: The workload executor did not write a results.json file in the expected location
 [2021/04/14 14:11:49.748] ----------------------------------------------------------------------
 [2021/04/14 14:11:49.748] Ran 7 tests in 32.554s
 [2021/04/14 14:11:49.748] FAILED (failures=4)
```

File: astrolabe/validator.py
Comment (jmikola) [reply astrolabe/validator.py:167]: Created [DRIVERS-1660](https://jira.mongodb.org/browse/DRIVERS-1660) to track this.

File: astrolabe/validator.py
Comment (p-mongo) [inline astrolabe/validator.py:96]: *is* permitted to?

**Review by p-mongo — APPROVED** [review]

**Comment by p-mongo** [comment]
After merging this PR can you please create a follow-up ticket to update the Ruby integration?

File: astrolabe/validator.py
Comment (alcaeus) [reply astrolabe/validator.py:96]:
```suggestion
        # The workload executor is permitted to exit prematurely when expecting
```

**Comment by jmikola** [comment]
> After merging this PR can you please create a follow-up ticket to update the Ruby integration?

Created [JAVA-4114](https://jira.mongodb.org/browse/JAVA-4114) and [RUBY-2587](https://jira.mongodb.org/browse/RUBY-2587).

**Pull request merged**
